src/Processors/Transforms/MergeJoinTransform.cpp

```diff
@@ -50,49 +50,17 @@ bool ALWAYS_INLINE isNullAt(const IColumn & column, size_t row)
     return false;
 }
 
-template <bool has_left_nulls, bool has_right_nulls>
+template <bool has_nulls>
 int nullableCompareAt(const IColumn & left_column, const IColumn & right_column, size_t lhs_pos, size_t rhs_pos, int null_direction_hint)
 {
-    if constexpr (has_left_nulls && has_right_nulls)
+    int res = left_column.compareAt(lhs_pos, rhs_pos, right_column, null_direction_hint);
+    if constexpr (has_nulls)
     {
-        const auto * left_nullable = checkAndGetColumn<ColumnNullable>(&left_column);
-        const auto * right_nullable = checkAndGetColumn<ColumnNullable>(&right_column);
-
-        if (left_nullable && right_nullable)
-        {
-            int res = left_nullable->compareAt(lhs_pos, rhs_pos, right_column, null_direction_hint);
-            if (res != 0)
-                return res;
-
-            /// NULL != NULL case
-            if (left_nullable->isNullAt(lhs_pos))
-                return null_direction_hint;
-
-            return 0;
-        }
+        /// NULL != NULL case
+        if (res == 0 && left_column.isNullAt(lhs_pos))
+            return null_direction_hint;
     }
-
-    if constexpr (has_left_nulls)
-    {
-        if (const auto * left_nullable = checkAndGetColumn<ColumnNullable>(&left_column))
-        {
-            if (left_nullable->isNullAt(lhs_pos))
-                return null_direction_hint;
-            return left_nullable->getNestedColumn().compareAt(lhs_pos, rhs_pos, right_column, null_direction_hint);
-        }
-    }
-
-    if constexpr (has_right_nulls)
-    {
-        if (const auto * right_nullable = checkAndGetColumn<ColumnNullable>(&right_column))
-        {
-            if (right_nullable->isNullAt(rhs_pos))
-                return -null_direction_hint;
-            return left_column.compareAt(lhs_pos, rhs_pos, right_nullable->getNestedColumn(), null_direction_hint);
-        }
-    }
```

Comment:
I'm not 100% sure, but it seems the logic has changed in a case when left column is not nullable and the right one is. Then this case would be applied in the old logic, and it's completely removed in your changes. Maybe it's expected, but I just want to be sure it's intended

Reply:
Initially this code was written under assumption that it's possible to have Nullable and non-Nullable columns on different sides. But for now it looks like we convert key columns to match types exactly https://github.com/ClickHouse/ClickHouse/blob/d99795b7636b6542a860ef2144283943d9f882f8/src/Planner/PlannerJoins.cpp#L843 so I decided ti update this code accordingly and verify on the CI if anything breaks or if we can simplify this code.

Reply:
Makes sense, I just suspected that something is odd looking that the fast test has failed and that's why I asked

Reply:
Seems key types can have different nullability with `join_use_nulls` ...